src/groth16/verifier.rs

```diff
@@ -53,3 +65,111 @@ pub fn verify_proof<'a, E: Engine>(
+    proofs: &[&Proof<E>],
+    public_inputs: &[Vec<E::Fr>],
+) -> Result<bool, SynthesisError>
+where
+    <<E as ff::ScalarEngine>::Fr as ff::PrimeField>::Repr: From<<E as ff::ScalarEngine>::Fr>,
+{
+    for pub_input in public_inputs {
+        if (pub_input.len() + 1) != pvk.ic.len() {
+            return Err(SynthesisError::MalformedVerifyingKey);
+        }
+    }
+
+    let pi_num = pvk.ic.len() - 1;
+    let proof_num = proofs.len();
+
+    // choose random coefficients for combining the proofs
+    let mut r: Vec<E::Fr> = Vec::with_capacity(proof_num);
+    for _ in 0..proof_num {
+        use rand::Rng;
+
+        let t: u128 = rng.gen();
+        let mut el = E::Fr::zero().into_repr();
+        let el_ref: &mut [u64] = el.as_mut();
+        assert!(el_ref.len() > 1);
+        el_ref[0] = (t & (-1i64 as u128) >> 64) as u64;
+        el_ref[1] = (t >> 64) as u64;
+
+        r.push(E::Fr::from_repr(el).unwrap());
+    }
+
+    let mut sum_r = E::Fr::zero();
+    for i in r.iter() {
+        sum_r.add_assign(i);
+    }
+
+    // create corresponding scalars for public input vk elements
+    let pi_scalars: Vec<_> = (0..pi_num)
+        .into_par_iter()
+        .map(|i| {
+            let mut pi = E::Fr::zero();
+            for j in 0..proof_num {
+                // z_j * a_j,i
+                let mut tmp = r[j];
+                tmp.mul_assign(&public_inputs[j][i]);
+                pi.add_assign(&tmp);
+            }
+            pi
+        })
+        .collect();
+
+    // create group element corresponding to public input combination
+    // This roughly corresponds to Accum_Gamma in spec
+    let mut acc_pi = pvk.ic[0].mul(sum_r.into_repr());
+    for (i, b) in pi_scalars.iter().zip(pvk.ic.iter().skip(1)) {
+        acc_pi.add_assign(&b.mul(i.into_repr()));
+    }
+
+    // This corresponds to Accum_Y
+    // -Accum_Y
+    sum_r.negate();
+    // This corresponds to Y^-Accum_Y
+    let acc_y = pvk.alpha_g1_beta_g2.pow(&sum_r.into_repr());
+
+    // This corresponds to Accum_Delta
+    let mut acc_c = E::G1::zero();
+    for (rand_coeff, proof) in r.iter().zip(proofs.iter()) {
+        let mut tmp: E::G1 = proof.c.into();
+        tmp.mul_assign(*rand_coeff);
+        acc_c.add_assign(&tmp);
+    }
+
+    // This corresponds to Accum_AB
+    let ml = r
+        .par_iter()
+        .zip(proofs.par_iter())
+        .map(|(rand_coeff, proof)| {
+            // [z_j] pi_j,A
+            let mut tmp: E::G1 = proof.a.into();
+            tmp.mul_assign(*rand_coeff);
+            let g1 = tmp.into_affine().prepare();
+
+            // -pi_j,B
+            let mut tmp: E::G2 = proof.b.into();
+            tmp.negate();
+            let g2 = tmp.into_affine().prepare();
+
+            (g1, g2)
+        })
+        .collect::<Vec<_>>();
+    let accum_ab_parts = ml.iter().map(|(a, b)| (a, b)).collect::<Vec<_>>();
+
+    // Accum_AB
+    let mut res = E::miller_loop(&accum_ab_parts);
+
+    res.mul_assign(&E::miller_loop(&[
```

Comment:
This will make code less conveniently match spec, but I guess instead of mul_assign we could push the two arguments below into accum_ab_parts and call miller_loop only once; might improve performance a bit